### DISCUSSION TITLE
Enhancing Directus Flows: Versioning, Archiving, and Improved Manageability

### DISCUSSION THREAD

**Discussion by kropsi** [directus/directus#19632] [Feature Requests]
### Summary

Enhancements in Directus Flows with features like versioning, archiving instead of deletion, security steps for deletion, and grouping or labeling for better organization and visibility.

### Basic Example

No change in API, but the UI should offer options for:
- Versioning Flows
- Archiving Flows instead of deletion
- Locking Flows to prevent accidental deletion
- Grouping or labeling Flows

### Motivation

The primary motivation is to make Flows more robust, secure, and manageable. The features suggested aim to:

1. Enable easy rollback to previous versions in case of failure (`Versioning`).
2. Prevent accidental loss of complex logic (`Archiving and Locking`).
3. Enhance manageability and findability (`Grouping/Labeling`).

These improvements will make Flows more viable for complex implementations, thereby increasing the utility and adoption of Directus.

#### Versioning:
- Allow users to save versions of a flow or its single operation configurations.
- Enable rollback to a previous version.

#### Archiving and Locking:
- Replace 'Delete' option with 'Archive'.
- Archived Flows can be restored.
- Implement a security step or a lock/unlock mechanism before deletion.

#### Grouping/Labeling:
- Add a feature to group or label Flows.
- Implement a search or filter functionality based on these groups or labels.

#### Must Have:
- Versioning of Flows
- Archiving instead of deleting Flows
- Security step for deleting Flows

#### Should Have:
- Ability to lock/unlock Flows
- Grouping or labeling of Flows

#### Could Have:
- Search or filter functionality based on groups or labels
- Notification system for version changes

#### Won't Have:
- Automatic deletion of old versions

### Drawbacks

- Increased complexity in UI/UX.
- Additional storage requirements for versioning.
- Possible learning curve for existing users to adapt to the new features.

### Alternatives

1. Manual backup of Flows, but this is not efficient and prone to errors.
2. Using third-party version control systems, which could be complex and costly.

### Adoption Strategy

Not a breaking change. Existing users will have their Flows intact but will notice the added features and functionalities. Documentation and tutorials can help in the transition.

### Unresolved Questions

- What would be the best method for implementing versioning?
- How much storage space will be required for archiving old versions?

**Comment by vwasteels** [comment]
at least a copy of the folder system of the data models would be great :)

**Comment by Pachat** [comment]
Copy a flow to start another with variations would help to avoid complete rebuilding.

**Comment by ChristopherJennings** [comment]
Update: [v12.4.1](https://github.com/directus/directus/releases/tag/v12.4.1) ships parts of this request (#28206).

Flows now have their own module, with a Manage Flows page that includes:

- **Folders** for grouping Flows
- **Search and filtering**
- **Duplication**, so you can copy a Flow and build a variation instead of starting over
- **Import and export**, which gives you a portable backup of a Flow before you make risky changes
